src/sdsc/xslt.py

```diff
@@ -0,0 +1,396 @@
+
+    if highlightstart >= len(tokens) or highlightend < highlightstart:
+        return ""  # Nothing to highlight
+
+    highlightend = min(highlightend, len(tokens) - 1)
+
+    tokens[highlightstart] = "<highlight>" + tokens[highlightstart]
+    tokens[highlightend] = tokens[highlightend] + "</highlight>"
+
+    return " ".join(tokens)
+
+
+def xmlescape(text):
+    """TODO@sknorr
+
+    :param str text:
+    """
+    escapetable = {
+        "&": "&amp;",
+        '"': "&quot;",
+        "'": "&apos;",
+        ">": "&gt;",
+        "<": "&lt;",
+        }
+    return "".join(escapetable.get(c, c) for c in text)
+
+
+def dupecheckmessage(context, quote, duplicate, contextid, basefile):
+    """TODO@sknorr"""
+    filename = ""
+    if basefile:
+        filename = "<file>%s</file>" % str(basefile)
+
+    withinid = ""
+    if contextid:
+        withinid = "<withinid>%s</withinid>" % str(contextid)
+
+    return etree.XML("""<result type="error">
+            <location>%s%s<line>%s</line></location>
+            <message><quote>%s</quote> is duplicated:
+                <quote>%s</quote>
+            </message>
+            <suggestion>Remove one instance of <quote>%s</quote>.</suggestion>
+        </result>""" % (filename, withinid, str(linenumber(context)), duplicate, quote, duplicate))
+
+
+# --------------------------------------------------------------------
+# XSLT Extension Functions
+# --------------------------------------------------------------------
+
+
+def linenumber(context):
+    """Returns the line number of the current context
+
+    :param str context: Context node (ignored)
+    """
+    return context.context_node.sourceline
+
+
+def splitpath(context, path, wantedsegment='filename'):
+    """Split a path into segments and return the wanted segment.
+
+    :param str context: Context node (ignored)
+    :param str path: Input path or file name
+    :param str wantedsegment: Segment that you want as output
+        ('path' => path without file name and without trailing slash,
+        'filename' => filename without extension,
+        'extension' => file extension with leading dot and in lowercase)
+    """
+
+    if not path:
+        return []
+
+    # I get this as a list with one lxml.etree._ElementUnicodeResult, not
+    # as a list with a string.
+    path = str(path[0])
+
+    splitpath = os.path.split(path)
+    splitfile = os.path.splitext(splitpath[1])
+
+    if wantedsegment == 'path':
+        return splitpath[0]
+    elif wantedsegment == 'filename':
+        return splitfile[0]
+    elif wantedsegment == 'extension':
+        return splitfile[1].lower().lstrip('.')
+    else:
+        return path
+
+
+def counttokens(context, text):
+    """TODO@sknorr:
+
+    :param str context: Context node (ignored)
+    """
+
+    count = 0
+    if text:
+        count = len(tokenizer(text[0]))
+    return count
+
+
+def tokenizer(text):
+    """TODO@sknorr: Splits text into words
```

Comment:
"Splits a string into a list of words"

Not much need for a fixme here, I guess..?